### PR TITLE
Introduces WfsSearch

### PR DESCRIPTION
**Pull request by KaiVolland**
### Description:

This introduces the `WfsSearch`.

### CODE REVIEW

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+     * The feature type names. Required.
+     */
+    featureTypes: PropTypes.arrayOf(PropTypes.string).isRequired,
+    /**
+     * SRS name. No srsName attribute will be set on geometries when this is not
+     * provided.
+     */
+    srsName: PropTypes.string,
+    /**
+     * Ther output format of the response.
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.jsx]: Typo `Ther` -> `The`

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+     */
+    className: PropTypes.string,
+    /**
+     * The base URL.
+     * @type {String}
+     */
+    baseUrl: PropTypes.string.isRequired,
+    /**
+     * The Property which should be shown in the inputfield when a selection was
+     * made.
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.jsx]: very minor: here and by some further props below `@type` annotation is missing

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    minChars: 3,
+    additionalFetchOptions: {},
+    /**
+     * Create an AutoComplete.Option from the given data.
+     *
+     * @param {Object} feature The feature as returned by the server.
+     * @return {AutoComplete.Option} The AutoComplete.Option that will be
+     * rendered for each feature.
+     */
+    renderOption: (feature) => {
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.jsx]: very minor: `()` around the argument can be ommited

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+        <Option key={feature.id}>
+          {feature.properties.name ? feature.properties.name : feature.id}
+        </Option>
+      );
+    },
+    /**
+     * The default onSelect method if no onSelect prop is given. It zooms to the
+     * selected item.
+     *
+     * @param {object} feature The selected feature as returned by the server.
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.jsx]: `olMap` param missing

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    const data = response.features ? response.features : [];
+    this.setState({
+      data,
+      fetching: false
+    });
+  }
+
+  /**
+   * This function gets called when the nomintim fetch returns an error.
+   * It logs the error to the console.
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.jsx]: Please adapt method description

File: src/Field/WfsSearch/WfsSearch.spec.jsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      //SETUP END
+
+      const wrapper = TestUtil.mountComponent(WfsSearch, {map});
+      const fitSpy = jest.spyOn(map.getView(), 'fit');
+      wrapper.props().onSelect(feature, map);
+      expect(fitSpy).toHaveBeenCalled();
+      setTimeout(() => {
+        expect(map.getView().getCenter()).toEqual([25, 25]);
+        expect(map.getView().getZoom()).toEqual(2);
+      }, 501);
```
Comment (annarieger) [inline src/Field/WfsSearch/WfsSearch.spec.jsx]: :sunglasses: :+1:

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+  createFilter() {
+    const {
+      searchTerm
+    } = this.state;
+    const {
+      searchAttributes
+    } = this.props;
+
+    const propertyFilters = searchAttributes.map(attribute =>
+      OlFormatFilter.like(attribute, `*${searchTerm}*`, '*', '.', '!', false));
```
Comment (ahennr) [inline src/Field/WfsSearch/WfsSearch.jsx]: What happens here if numerical / boolean attributes are filtered / queried here?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+      featureTypes,
+      geometryName,
+      maxFeatures,
+      outputFormat,
+      propertyNames,
+      srsName,
+      filter: this.createFilter()
+    };
+
+    const wfsFormat = new OlFormatWFS();
```
Comment (ahennr) [inline src/Field/WfsSearch/WfsSearch.jsx]: By default `OlFormatWFS` supports WFS version 1.1. Please make the used WFS version configurable (see e.g. [here](https://openlayers.org/en/latest/apidoc/ol.format.WFS.html)).

WFS 2.0.x is not supported, or am I wrong?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    const featureRequest = wfsFormat.writeGetFeature(options);
+    this.setState({fetching: true});
+
+    fetch(`${baseUrl}`, {
+      method: 'POST',
+      credentials: additionalFetchOptions.credentials
+        ? additionalFetchOptions.credentials
+        : 'same-origin',
+      body: new XMLSerializer().serializeToString(featureRequest),
```
Comment (ahennr) [inline src/Field/WfsSearch/WfsSearch.jsx]: I've in mind that there existed some problems using `XMLSerializer` in Firefox.

Did you check this also other browsers like FF, Edge, ..?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  static propTypes = {
+    /**
+     * A optional className that will be added to the input field.
+     */
+    className: PropTypes.string,
+    /**
+     * The base URL.
+     * @type {String}
+     */
+    baseUrl: PropTypes.string.isRequired,
```
Comment (ahennr) [inline src/Field/WfsSearch/WfsSearch.jsx]: Must CORS headers be set by the WFS server?

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 7 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: Uff i don't know. I think this depends on the WFS-Server. Either it throws an error or it just won't match... Anything we can do about it?

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 8 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: `ol.format.WFS` does not support WFS 2.0.0, is it really necessary to support WFS 1.0.0?

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 9 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: No i didn't nice hint. I'll check this.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 10 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: Yes, i think so.

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+     * The Property which should be shown in the inputfield when a selection was
+     * made.
+     * @type {String}
+     */
+    displayField: PropTypes.string,
+    /**
+     * The list of attributes that should be searched through.
+     * @type {String[]}
+     */
+    searchAttributes: PropTypes.arrayOf(PropTypes.string).isRequired,
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: Can we get a utility to get these from a `DescribeFeatureType`? Posibbly first fetching these and filtering to those usable?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+     * The base URL.
+     * @type {String}
+     */
+    baseUrl: PropTypes.string.isRequired,
+    /**
+     * The Property which should be shown in the inputfield when a selection was
+     * made.
+     * @type {String}
+     */
+    displayField: PropTypes.string,
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: Can we change this to be a template or a function, so I can show multiple fields and arbitrary text? Also it would be super nice if we could somehow mark the matched textpart in the property.

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+    /**
+     * Maximum number of features to fetch.
+     * @type {number}
+     */
+    maxFeatures: PropTypes.number,
+    /**
+     * Geometry name to use in a BBOX filter.
+     * @type {string}
+     */
+    geometryName: PropTypes.string,
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: A utility to get this from a DescribeFeatureType would be nice

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+     * for more information.
+     * @type {object}
+     */
+    filter: PropTypes.object,
+    /**
+     * The ol.map where the map will zoom to.
+     *
+     * @type {Object}
+     */
+    map: PropTypes.instanceOf(OlMap).isRequired,
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: Why is this required? we can search without a map I guess?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+    minChars: 3,
+    additionalFetchOptions: {},
+    /**
+     * Create an AutoComplete.Option from the given data.
+     *
+     * @param {Object} feature The feature as returned by the server.
+     * @return {AutoComplete.Option} The AutoComplete.Option that will be
+     * rendered for each feature.
+     */
+    renderOption: feature => {
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: If a user can provide this, we should list actual requirements, if any. I am thinking of the id property, e.g.

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+      if (feature) {
+        const olView = olMap.getView();
+        const geoJsonFormat = new OlFormatGeoJSON();
+        const olFeature = geoJsonFormat.readFeature(feature);
+        const geometry = olFeature.getGeometry();
+
+        if (geometry) {
+          olView.fit(geometry, {
+            duration: 500
+          });
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: These options might be a good candidate for configuration, right?

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+   *                                      is pressed.
+   */
+  onUpdateInput(value) {
+    this.setState({
+      data: []
+    });
+
+    if (value) {
+      this.setState({
+        searchTerm: value || ''
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.jsx]: `|| ''` seems unneeded

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 8 above)
Comment (marcjansen) [reply src/Field/WfsSearch/WfsSearch.jsx]: no need for 1.0.0 IMO. 2.0.0 would be nice, but is not a must have.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 7 above)
Comment (marcjansen) [reply src/Field/WfsSearch/WfsSearch.jsx]: for geoserver we can cast e.g. by using a string function like concatenate with an empty string, which will accept any input (according to the docs). this should be configurable defaulting to false probably.

File: src/Field/WfsSearch/WfsSearch.spec.jsx [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+      });
+      //SETUP END
+
+      const wrapper = TestUtil.mountComponent(WfsSearch, {map});
+      const fitSpy = jest.spyOn(map.getView(), 'fit');
+      wrapper.props().onSelect(feature, map);
+      expect(fitSpy).toHaveBeenCalled();
+      setTimeout(() => {
+        expect(map.getView().getCenter()).toEqual([25, 25]);
+        expect(map.getView().getZoom()).toEqual(2);
```
Comment (marcjansen) [inline src/Field/WfsSearch/WfsSearch.spec.jsx]: is this actually tested?

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 11 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: Sounds very useful. Can you open an issue for it.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 13 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: Sounds very useful. Can you open an issue for it.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 12 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: After having a look at the code i found out, that this prop is never used. I'll remove it. All what you expect should be realized in the `renderOption` prop.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 15 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: I'll enhance the docs at the property.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 16 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: As it's just in the default of the `onSelect` property this doesn't need to be configurable. You can just provide a complete new function if you want a different duration.

File: src/Field/WfsSearch/WfsSearch.spec.jsx (hunk 18 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.spec.jsx]: `expect.assertions` is missing. I'll add it.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 9 above)
Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: After a quick check, it works as aspected in FireFox and InternetExpolrer.

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 8 above)
Comment (ahennr) [reply src/Field/WfsSearch/WfsSearch.jsx]: Since WFS 1.0.0 will be generated simply by using another GML format in `OlFormatWFS` I would make this configurable

Comment (KaiVolland) [reply src/Field/WfsSearch/WfsSearch.jsx]: I :womans_hat: this done in 638c12d

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 11 above)
Comment (marcjansen) [reply src/Field/WfsSearch/WfsSearch.jsx]: Here you are: #567

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 13 above)
Comment (marcjansen) [reply src/Field/WfsSearch/WfsSearch.jsx]: Here you are: #567
